src/criticality/BoratedWater.C

```diff
@@ -0,0 +1,223 @@
+  // any nuclides there to be H2O + boron weight ppm and add the S(a,b) tables
+  // for hydrogen. Therefore, we perform a check here if any nuclides in the
+  // material are *not* those we are going to add back when we wipe the material,
+  // to notify the user they probably provided the wrong material ID or they need
+  // to use a more general material modification object that allows full control
+  // over additional nuclides (e.g., if their water includes corrosion products
+  // that they do want to be there).
+
+  const int * nuclides;
+  const double * densities;
+  int n;
+  int err = openmc_material_get_densities(_material_index, &nuclides, &densities, &n);
+  catchOpenMCError(err, "get nuclide densities from material " + std::to_string(_material_id));
+
+  // get all the natural isotopes of H, O, B
+  _hydrogen_natural = NuclearData::Nuclide::getAbundances("H");
+  _boron_natural = NuclearData::Nuclide::getAbundances("B");
+  _oxygen_natural = NuclearData::Nuclide::getAbundances("O");
+
+  std::vector<std::string> allowable;
+  for (const auto & i : _hydrogen_natural)
+    allowable.push_back(i.first);
+  for (const auto & i : _boron_natural)
+    allowable.push_back(i.first);
+  for (const auto & i : _oxygen_natural)
+    allowable.push_back(i.first);
+
+  if (isParamValid("absent_nuclides"))
+  {
+    const auto & absent = getParam<std::vector<std::string>>("absent_nuclides");
+
+    for (const auto & a : absent)
+    {
+      // only missing nuclides for H, O, B are meaningful
+      if (std::find(allowable.begin(), allowable.end(), a) == allowable.end())
+        paramWarning("absent_nuclides",
+                     "Only absent isotopes of hydrogen, oxygen, or boron will be used to adjust "
+                     "natural abundances. The entry '" +
+                         a + "' will be unused.");
+
+      // adjust the natural abundances if nuclides are missing from the cross section library
+      // TODO: implement in a general fashion, this only works for O18 which is also the only
+      // absent nuclide we expect in practice
+      if (a == "O18")
+      {
+        // find which entry in _oxygen_natural has O16 and O18; this allows these isotopes
+        // to appear in any order
+        unsigned int idx16;
+        unsigned int idx18;
+        unsigned int idx = 0;
+        for (const auto & o : _oxygen_natural)
+        {
+          if (o.first == "O16")
+            idx16 = idx;
+          idx++;
+        }
+
+        idx = 0;
+        for (const auto & o : _oxygen_natural)
+        {
+          if (o.first == "O18")
+            idx18 = idx;
+          idx++;
+        }
+
+        _oxygen_natural[idx16].second += _oxygen_natural[idx18].second;
+        _oxygen_natural.erase(_oxygen_natural.begin() + idx18);
+      }
+      else
+        paramError(
+            "absent_nuclides",
+            "Cardinal currently only assumes that O18 may be missing from your cross section "
```

Comment:
All documentation and code comments suggest that H and B are handled also, but this message contradicts this.  Should the docs be consistent with this?

Reply:
I see both perspectives on this -- my philosophy has been for the description of input parameters to reflect the ideal, end-code state and to handle any gaps using error messages. For example, there is a lot of this which happens using enumerations for types of data transfers, only some of which are supported by class A, class B, etc. Rather than define separate enumerations for every class having to do with data transfer, I simply re-use the same enumeration and have error messages. 

Otherwise, I think a better name for this parameter would be `O18_missing` (bool), but I prefer to add the parameter `absent_nuclides` to allow easy extension in the future while supporting backwards compatibility.

But it's not a big deal to modify this - let me know if you prefer that.